base/acme/src/main/java/org/dogtagpki/acme/validator/HTTP01Validator.java

```diff
@@ -47,49 +47,46 @@ public void validateChallenge(
         ACMEIdentifier identifier = authorization.getIdentifier();
         String hostname = identifier.getValue();
         String validationPath = "/.well-known/acme-challenge/" + token;
-        URI validationURL = new URI("http", hostname, validationPath, null);
-
-        // TODO: move retry to ACMEChallengeProcessor.processChallenge()
-        // TODO: make it configurable
-
-        int maxCount = 5;
-        int interval = 5;
-
-        int count = 1;
-        String response = null;
-        Exception exception = null;
-
-        while (true) {
-            try {
-                response = getResponse(validationURL);
-                break;
-
-            } catch (Exception e) {
-
-                // TODO: catch more specific HTTP exception
 
-                logger.error(e.getMessage());
-                if (count >= maxCount) {
-                    exception = e;
-                    break;
-                }
-
-                Thread.sleep(interval * 1000);
-                count++;
-            }
+        URI validationURL;
+        try {
+            /* It was suggested to catch URISyntaxException and return
+             * urn:ietf:params:acme:error:malformed.  A close inspection leads
+             * to a different conclusion.
+             *
+             * `hostname` comes from `identifier` object, via `authorization`.
+             * `validationPath` is derived from `token` which comes from
+             * `challenge`.  All of the source objects are looked up
+             * server-side, not supplied by the client.  If we got this far,
+             * the client request was fine but there are missing or corrupt
+             * data on the server side.  Therefore the correct error is
+             * urn:ietf:params:acme:error:serverInternal.
+             */
+            validationURL = new URI("http", hostname, validationPath, null);
+        } catch (URISyntaxException e) {
```

Comment:
According to the javadoc, the `URISyntaxException` will be thrown if it violates RFC 2396 (which defines hostname syntax as well):
https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#URI-java.lang.String-java.lang.String-java.lang.String-java.lang.String-

So maybe we should just use `new URI()` to validate the hostname in `ACMENewOrderService` and again here to make the actual HTTP connection? Since we use the same code, we can be certain that the `new URI()` here will not generate an `URISyntaxException` since it's already validated earlier.

Reply:
Hm, well, the URI we build in `HTTP01Validator` includes the `validationPath`, which we do not know at the time when we validate the identifer in `ACMENewOrderService`.   We can only check that the URI `https://{hostname}` is valid.  If the value of `validationPath` somehow makes for an invalid URI then we still get `URISyntaxException` (and it is indeed an `serverInternal` error, because the server decides the `validationPath`.

That said, it is a good idea to construct a `new URI()` with the hostname as an additional check when validating the identifier syntax, and I will implement that.